packages/alfa-rules/src/sia-r45/rule.ts

```diff
@@ -0,0 +1,95 @@
+import { Rule } from "@siteimprove/alfa-act";
+import { Attribute, Element, Namespace } from "@siteimprove/alfa-dom";
+import { Map } from "@siteimprove/alfa-map";
+import { None, Option, Some } from "@siteimprove/alfa-option";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Err, Ok } from "@siteimprove/alfa-result";
+import { Page } from "@siteimprove/alfa-web";
+import { expectation } from "../common/expectation";
+import { hasAttribute } from "../common/predicate/has-attribute";
+import { isPerceivable } from "../common/predicate/is-perceivable";
+
+const { isElement, hasName, hasNamespace } = Element;
+const { and, not } = Predicate;
+
+export default Rule.Atomic.of<Page, Attribute>({
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r45.html",
+  evaluate({ device, document }) {
+    // records in which <table> is located each "headers" attribute
+    let ownership: Map<Attribute, Element> = Map.empty();
+
+    return {
+      applicability() {
+        return (
+          document
+            .descendants()
+            // get all cells with a headers attribute
+            .filter(
+              and(
+                isElement,
+                and(
+                  hasNamespace(Namespace.HTML),
+                  hasName("td", "th"),
+                  hasAttribute("headers")
+                )
+              )
+            )
+            // get the attributes themselves
+            .map((cell) => {
+              let result: Option<Attribute> = None;
+              // get the table containing it: the first <table> ancestor
+              const table = cell.closest(
+                and(
+                  isElement,
+                  and(hasNamespace(Namespace.HTML), hasName("table"))
+                )
+              );
+
+              if (table.isNone()) return result;
+              if (not(isPerceivable(device))(table.get())) return result;
+
+              // if the table is exposed, record the ownership and return the headers attribute
+              const headers = cell.attribute("headers").get();
+              ownership = ownership.set(headers, table.get());
+              result = Some.of(headers);
+              return result;
+            })
+            .filter((option) => option.isSome())
+            .map((option) => option.get())
```

Comment:
A cleaner approach to this would probably be to start from each `<table>` and then drill down into the cells, rather than the other way around. It's becomes a little awkward having to repeatedly walk back up the tree to look for the owning table, and repeatedly checking if the same table is perceivable.